plutus-tx/src/PlutusTx/Data/List.hs

```diff
@@ -0,0 +1,194 @@
+                         pred)
+import Prettyprinter (Pretty (..))
+
+import Data.Semigroup qualified as Haskell
+import Prelude qualified as Haskell
+
+-- | A list type backed directly by 'Data'. It is meant to be used whenever fast
+-- encoding/decoding to/from 'Data' is needed.
+newtype List a = List (BuiltinList BuiltinData)
+  deriving stock (Haskell.Show, Haskell.Eq)
+
+instance Eq (List a) where
+    {-# INLINEABLE (==) #-}
+    List l == List l' =
+        case (B.uncons l, B.uncons l') of
+            (Just (h, t), Just (h', t')) -> h == h' && List t == List t'
+            (Nothing, Nothing)           -> True
+            _                            -> False
+
+instance ToData (List a) where
+    {-# INLINEABLE toBuiltinData #-}
+    toBuiltinData (List l) = BI.mkList l
+
+instance FromData (List a) where
+    {-# INLINEABLE fromBuiltinData #-}
+    fromBuiltinData = Just . List . BI.unsafeDataAsList
+
+instance UnsafeFromData (List a) where
+    {-# INLINEABLE unsafeFromBuiltinData #-}
+    unsafeFromBuiltinData = List . BI.unsafeDataAsList
+
+instance (UnsafeFromData a, Pretty a) => Pretty (List a) where
+    {-# INLINEABLE pretty #-}
+    pretty = pretty . toSOP
+
+append :: List a -> List a -> List a
+append (List l) (List l') = List (go l l')
+  where
+    go xs ys =
+        B.caseList'
+            ys
+            (\h t -> BI.mkCons h (go t ys))
+            xs
+
+instance Semigroup (List a) where
+    (<>) = append
+
+instance Monoid (List a) where
+    mempty = List B.mkNil
+
+instance Haskell.Semigroup  (List a) where
+    (<>) = append
+
+instance Haskell.Monoid (List a) where
+    mempty = List B.mkNil
+
+{-# INLINEABLE toSOP #-}
+toSOP :: (UnsafeFromData a) => List a -> [a]
+toSOP (List l) = go l
+  where
+    go = B.caseList' [] (\h t -> unsafeFromBuiltinData h : go t)
+
+{-# INLINEABLE fromSOP #-}
+fromSOP :: (ToData a) => [a] -> List a
+fromSOP = List . BI.unsafeDataAsList . B.mkList . fmap toBuiltinData
+
+{-# INLINEABLE find #-}
+find :: (UnsafeFromData a) => (a -> Bool) -> List a -> Maybe a
+find pred (List l) = go l
+  where
+    go =
+        B.caseList'
+            Nothing
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                in
+                    if pred h'
+                        then Just h'
+                        else go t
+            )
+
+-- TODO: return [] or List?
+{-# INLINEABLE findIndices #-}
+findIndices :: (UnsafeFromData a) => (a -> Bool) -> List a -> [Integer]
+findIndices pred (List l) = go 0 l
+  where
+    go i =
+        B.caseList'
+            []
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                    indices = go (B.addInteger i 1) t
+                in
+                    if pred h'
+                        then i : indices
+                        else indices
+            )
+
+-- TODO: return [] or List?
+{-# INLINEABLE filter #-}
+filter :: (UnsafeFromData a) => (a -> Bool) -> List a -> [a]
```

Comment:
I think `[]`, because what's the point of converting the elements from `Data` just to throw them away. But that does suggest that we should have two versions for each such function: one operating on `Data` like

```haskell
filterData :: (Data -> Bool) -> List a -> List a
```

and one as it is currently defined.

So I think the answer to "return `[]` or `List`?" is "yes" here.

Reply:
> Converting elements to/from `Data` is supposed to be extremely cheap, since the `a` in `List a` is supposed to be `Data`-encoded

It is `Data`-encoded, but what you do here is decode it fully to the type that is represented by `Data`, which for all we know here (since this module is a general one and isn't tied to the `Data` version of `ScriptContext`) can be arbitrarily complex. And then you throw all of that work away.